collector/container.go

```diff
@@ -0,0 +1,258 @@
+// returns containers metrics from HCS
+// https://github.com/Microsoft/hcsshim
+
+// +build windows
+
+package collector
+
+import (
+	"github.com/Microsoft/hcsshim"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+func init() {
+	Factories["container"] = NewContainerMetricsCollector
+}
+
+// A ContainerMetricsCollector is a Prometheus collector for containers metrics
+type ContainerMetricsCollector struct {
+	// Presence
+	ContainerAvailable *prometheus.Desc
+	// memory
+	UsageCommitBytes            *prometheus.Desc
+	UsageCommitPeakBytes        *prometheus.Desc
+	UsagePrivateWorkingSetBytes *prometheus.Desc
+
+	// CPU
+	TotalRuntime100ns  *prometheus.Desc
+	RuntimeUser100ns   *prometheus.Desc
+	RuntimeKernel100ns *prometheus.Desc
+
+	// Network
+	BytesReceived          *prometheus.Desc
+	BytesSent              *prometheus.Desc
+	PacketsReceived        *prometheus.Desc
+	PacketsSent            *prometheus.Desc
+	DroppedPacketsIncoming *prometheus.Desc
+	DroppedPacketsOutgoing *prometheus.Desc
+}
+
+func NewContainerMetricsCollector() (Collector, error) {
+	const subsystem = "container"
+	return &ContainerMetricsCollector{
+		ContainerAvailable: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "available"),
+			"Available",
+			[]string{"container_id"},
+			nil,
+		),
+		UsageCommitBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "memory_usage_commit_bytes"),
+			"Memory Usage Commit Bytes",
+			[]string{"container_id"},
+			nil,
+		),
+		UsageCommitPeakBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "memory_usage_commit_peak_bytes"),
+			"Memory Usage Commit Peak Bytes",
+			[]string{"container_id"},
+			nil,
+		),
+		UsagePrivateWorkingSetBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "memory_usage_private_working_set_bytes"),
+			"Memory Usage Private Working Set Bytes",
+			[]string{"container_id"},
+			nil,
+		),
+		TotalRuntime100ns: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "cpu_usage_seconds_total"),
```

Comment:
Does usermode and kernelmode add upp to total, or is there some "other" time too? It might be more in line with how we report other cpu usage to have a single metric with a `mode` label, if possible.

Reply:
I have to confirm with the folks who wrote the hcsshim layer. Though as we have only 2 modes so having separate metrics for them should not be big problem.